funsor/tensor.py

```diff
@@ -88,6 +88,16 @@ def __call__(cls, data, inputs=None, dtype="real"):
         # it seems that there is no harm with the conversion generic -> ndarray here
         if isinstance(data, np.generic):
             data = data.__array__()
+        # XXX: in jaxlib 0.1.58 + jax 0.2.8, sometime opt_einsum returns
+        # a jaxlib.xla_client.Buffer, which is a type not supported in Tensor.
+        # It is unclear whether this is an issue. The good thing is: under jit,
+        # data will be a ShapedArray, so we won't go to this branch to do
+        # the extra job `jax.numpy.array(data)`.
+        elif "Buffer" in type(data).__name__ and get_backend() == "jax":
+            import jax
+
+            data = jax.numpy.array(data)
```

Comment:
Do we need to add something like this to `funsor.jax.ops.einsum` as well?

Reply:
@hawkinsp In https://github.com/google/jax/pull/2839, you mentioned that `xla_client.Buffer` is going to be deleted. But in the latest jax+jaxlib release, we are seeing that `Buffer` starts appearing (it didn't appear in previous releases). I just wonder if it is will be supported in the long run?

Reply:
Ah, this is a different `Buffer`. The one you are talking about was a Python class. The new one is a C++ class.

To improve dispatch times, we are moving a few parts of the JAX dispatch logic into C++, and in some cases (notably, uncomplicated `jit` calls), you will get a pure C++ object out (`Buffer`) without the additional Python wrapper class. It should duck type the same (e.g., have the same methods and act the same in most respects).

I believe both the Python and C++ variants report themselves as being instances of `DeviceArray`. Eventually we will only have the C++ version.

Reply:
Thanks for the very clear explanation, @hawkinsp! I understand what's going on now.